tests/push/test_push_rules_http.py

```diff
@@ -0,0 +1,117 @@
+        login.register_servlets,
+        receipts.register_servlets,
+        push_rule.register_servlets,
+    ]
+    user_id = True
+    hijack_auth = False
+
+    def default_config(self):
+        config = super().default_config()
+        config["start_pushers"] = True
+        return config
+
+    def make_homeserver(self, reactor, clock):
+        self.push_attempts: List[tuple[Deferred, str, dict]] = []
+
+        m = Mock()
+
+        def post_json_get_json(url, body):
+            d: Deferred = Deferred()
+            self.push_attempts.append((d, url, body))
+            return make_deferred_yieldable(d)
+
+        m.post_json_get_json = post_json_get_json
+
+        hs = self.setup_test_homeserver(proxied_blacklisted_http_client=m)
+
+        return hs
+
+    def _make_user_with_pusher(self, username):
+        user_id = self.register_user(username, "pass")
+        access_token = self.login(username, "pass")
+
+        # Register the pusher
+        user_tuple = self.get_success(
+            self.hs.get_datastores().main.get_user_by_access_token(access_token)
+        )
+        token_id = user_tuple.token_id
+
+        self.get_success(
+            self.hs.get_pusherpool().add_pusher(
+                user_id=user_id,
+                access_token=token_id,
+                kind="http",
+                app_id="m.http",
+                app_display_name="HTTP Push Notifications",
+                device_display_name="pushy push",
+                pushkey="a@example.com",
+                lang=None,
+                data={"url": "http://example.com/_matrix/push/v1/notify"},
+            )
+        )
+
+        return user_id, access_token
+
+    def test_dont_notify_rule_overrides_message(self):
+        """
+        The override push rule will suppress notification
+        """
+
+        user_id, access_token = self._make_user_with_pusher("user")
+        other_user_id, other_access_token = self._make_user_with_pusher("otheruser")
+
+        # Create a room
+        room = self.helper.create_room_as(user_id, tok=access_token)
+
+        # Disable user notifications for this room -> user
+        body = {
+            "conditions": [{"kind": "event_match", "key": "room_id", "pattern": room}],
+            "actions": ["dont_notify"],
+        }
+        channel = self.make_request(
+            "PUT",
+            "/pushrules/global/override/best.friend",
+            body,
+            access_token=access_token,
+        )
+        self.assertEqual(channel.code, 200)
+
+        # The other user joins
+        self.helper.join(room=room, user=other_user_id, tok=other_access_token)
+
+        # The other user sends a message (ignored by dont_notify push rule set above)
+        self.helper.send(room, body="Hi!", tok=other_access_token)
+
+        # The user sends a message back (sends a notification)
+        self.helper.send(room, body="Hello", tok=access_token)
+
+        self.assertEqual(len(self.push_attempts), 1)
```

Comment:
We should probably assert that it is 0 after the first message is sent, so we know which of the above messages is causing the push:

```suggestion
        # The other user sends a message (ignored by dont_notify push rule set above)
        self.helper.send(room, body="Hi!", tok=other_access_token)
        self.assertEqual(len(self.push_attempts), 0)

        # The user sends a message back (sends a notification)
        self.helper.send(room, body="Hello", tok=access_token)
        self.assertEqual(len(self.push_attempts), 1)
```

(We even might want to do it before sending the first event too?)

Reply:
https://github.com/matrix-org/synapse/pull/12188/commits/39b8758777a6cde5d34b74072a36a7deaa59ea73